src/common/meta/src/ddl/truncate_table.rs

```diff
@@ -0,0 +1,233 @@
+            table_ref.schema,
+            table_ref.table,
+        );
+
+        LockKey::single(key)
+    }
+}
+
+impl TruncateTableProcedure {
+    pub(crate) const TYPE_NAME: &'static str = "metasrv-procedure::TruncateTable";
+
+    pub(crate) fn new(
+        cluster_id: u64,
+        task: TruncateTableTask,
+        table_info_value: TableInfoValue,
+        region_routes: Vec<RegionRoute>,
+        context: DdlContext,
+    ) -> Self {
+        Self {
+            context,
+            data: TruncateTableData::new(cluster_id, task, table_info_value, region_routes),
+        }
+    }
+
+    pub(crate) fn from_json(json: &str, context: DdlContext) -> ProcedureResult<Self> {
+        let data = serde_json::from_str(json).context(FromJsonSnafu)?;
+        Ok(Self { context, data })
+    }
+
+    // Checks whether the table exists.
+    async fn on_prepare(&mut self) -> Result<Status> {
+        let table_ref = &self.data.table_ref();
+
+        let manager = &self.context.table_metadata_manager;
+
+        let exist = manager
+            .table_name_manager()
+            .exists(TableNameKey::new(
+                table_ref.catalog,
+                table_ref.schema,
+                table_ref.table,
+            ))
+            .await?;
+
+        ensure!(
+            exist,
+            TableNotFoundSnafu {
+                table_name: table_ref.to_string()
+            }
+        );
+
+        self.data.state = TruncateTableState::DatanodeTruncateRegions;
+
+        Ok(Status::executing(true))
+    }
+
+    async fn on_datanode_truncate_regions(&mut self) -> Result<Status> {
+        let table_id = self.data.table_id();
+
+        let region_routes = &self.data.region_routes;
+        let leaders = find_leaders(region_routes);
+        let mut truncate_region_tasks = Vec::with_capacity(leaders.len());
+
+        for datanode in leaders {
+            let clients = self.context.datanode_manager.clone();
+            let regions = find_leader_regions(region_routes, &datanode);
+            let region_ids = regions
+                .iter()
+                .map(|region_number| RegionId::new(table_id, *region_number))
+                .collect::<Vec<_>>();
+
+            truncate_region_tasks.push(async move {
+                for region_id in region_ids {
+                    debug!("Truncating region {region_id} on Datanode {datanode:?}");
+
+                    let request = RegionRequest {
+                        header: Some(RegionRequestHeader {
+                            trace_id: 0,
+                            span_id: 0,
+                        }),
+                        body: Some(region_request::Body::Truncate(PbTruncateRegionRequest {
+                            region_id: region_id.as_u64(),
+                        })),
+                    };
+
+                    if let Err(err) = clients.datanode(&datanode).await.handle(request).await {
+                        if err.status_code() != StatusCode::RegionNotFound {
```

Comment:
I don't think we will incur `RegionNotFound`. Any procedures executed in `metasrv`, acquire a table name lock.

Reply:
Yes, you are right.